# Lab 2

## Imports

### Documentation Stage 1

For this lab, there is an assortment of necessary libraries: 
- itertools for generating dice combinations.
- numpy for efficient calculations and array manipulation assistance.
- pandas for data manipulation.


In [8]:
import itertools
import pandas as pd
import numpy as np

## Function A

### Documentation Stage 2

We need a function to calculate the probability of getting a sum greater than x when rolling m dice with n sides.
The reason that this function is necessary is so that players are able to gauge their option and find out the true risk to an action.

The function is defined as below, using the 3 arguments 'n', 'm', and 'x' to deliniate the dice side amount, the number of dice, and the target sum respectively.

In [9]:
def m_dice_n_sides_prob_sum_x(n: int, m: int, x: int) -> float:
    """
    Calculate the probability of getting a sum greater than x when rolling m dice with n sides.
    n: number of sides of the dice. Int from 1 to 20.
    m: number of dice. Int from 1 to 4.
    x: target sum.
    """
    if n < 1 or n > 20 or m < 1 or m > 4:
        # These are the limits of the problem. If we go beyond these, we must throw an error.
        raise ValueError("n must be an integer from 1 to 20 and m must be an integer from 1 to 4.")
    # List out all possible rolls. It will basically be a list of tuples where each tuple is a roll.
    # For example, for m = 2 and n = 6, the list will be [(1, 1), (1, 2), ..., (6, 6)]
    rolls = list(itertools.product(range(1, n + 1), repeat=m))
    # Find the sum of all possible rolls. This will be a list of sums of each roll.
    # In the example above, it will be [2, 3, ..., 12]
    sums = np.array([sum(roll) for roll in rolls])
    # Find the proportion of those sums that are greater than the target x
    prob = np.mean(sums > x)
    return float(prob)

In [10]:
m_dice_n_sides_prob_sum_x(6, 3, 11)

0.375

## Function B

### Documentation Stage 3

Another necessary function is that of having at least 1 dice be larger than a required value. This sort of task is extremely hard to think through mentally, allowing the power of software computing to shine. This is useful in the art of Dungeon and Dragons, since there are many instances in which the game master may use probability of rolling a die to determine the outcome of player actions, such as avoiding traps, making critical hits, or succeeding in skill checks.

The function is defined as below, using the 3 arguments 'n', 'm', and 'x' to deliniate the dice side amount, the number of dice, and the target value for a single dice to hit respectively.

In [11]:
def prob_at_least_one_greater_equal_x(n: int, m: int, x: int) -> float:
    """
    Determine the probability of rolling m dice, each with n sides, and having at least one die
    show a value greater than or equal to x.
    n: number of sides of the dice.
    m: number of dice.
    x: target value.
    """
    # List of all possible dice rolls
    rolls = list(itertools.product(range(1, n + 1), repeat=m))
    # Calculate how many rolls have at least one die >= x
    success_rolls = [roll for roll in rolls if any(r >= x for r in roll)]
    # Probability calculation
    return len(success_rolls) / len(rolls)

In [12]:
# Probability the player will roll at least one 5 or 6 when rolling 2d6
prob_at_least_one_greater_equal_x(6, 2, 5)

0.5555555555555556

## WALKTHROUGH

A player is in a challenging situation where they have a choice. They can choose a single challenge from the following pool of challenges:

- getting a sum of 17 or more with 1d20
- getting a sum of 10 or more with 3d4
- getting at least one dice roll of 6 with 4d6

Which one is the best for the player? That is the question we will be answering below.

In [13]:
# 1d20 and get a result of 17 or more
# The function is called with 16 because the function calculates the probability of getting a sum greater than x
challenge_1 = m_dice_n_sides_prob_sum_x(20, 1, 16)
# 3d4 and get a result of 10 or more
# The function is called with 9 because the function calculates the probability of getting a sum greater than x
challenge_2 = m_dice_n_sides_prob_sum_x(4, 3, 9)
# 4d6 and have at least one die show a 6
challenge_3 = prob_at_least_one_greater_equal_x(6, 4, 6)

# Output the results with 4 decimal places
print(f"Probability of getting a sum of 17 or more with 1d20: {challenge_1:.4f}")
print(f"Probability of getting a sum of 10 with 3d4: {challenge_2:.4f}")
print(f"Probability of getting at least one dice roll of 6 with 4d6: {challenge_3:.4f}")

Probability of getting a sum of 17 or more with 1d20: 0.2000
Probability of getting a sum of 10 with 3d4: 0.1562
Probability of getting at least one dice roll of 6 with 4d6: 0.5177


### Walkthrough Conclusion

From the probabilities, it seems that challenge 3 *(getting at least one dice to roll a 6 from an assortment of 4 d6's)* is **by far** the best choice. It is to the point that you could try the other options **many times over** and they still would not be a better choice than challenge 3. With a **~52%** probability of success, I am willing to take my chances.

The second pick for challenges would be challenge 1. Although it is still really low with a **20%** probability of success, I would take it any day over challenge 2.

Challenge 2 is a b-line to death row. **~15%** chance of success? Only the most adventurous of people would be willing to take this challenge over the other 2.